jep/0000/README.adoc

```diff
@@ -0,0 +1,225 @@
+
+However:
+
+* Various plugins call `Run.getArtifactsDir` and similar deprecated APIs.
+These would already have been broken for users of the Compress Artifacts plugin, but that is rarely used, whereas we are proposing lots of people run with the S3 artifact manager.
+Calls to the deprecated APIs will behave as if there were no artifacts in the build.
+We could add telemetry so that such calls produce a warning in the system log, at least when the build actually does have a custom artifact manager selected.
+
+* Some plugins using `VirtualFile` may still be calling open and then passing the stream to an agent.
+This will work, but will be very expensive when using S3 storage. They need to be updated to call `VirtualFile.toExternalURL`.
+
+== Security
+
+Security considerations make agents need to be restricted to only access the artifacts needed. Having access to the blob store would mean access to other jobs artifacts. Two possible implementations were considered:
+
+=== Agents only need upload/download permissions
+
+If agents only do upload/download operations we can use pre-signed urls so they will not be able to access other jobs artifacts.
+Other operations (list, create, delete,...) would run on the master, which would be a performance hit for builds with many artifacts
+
+=== Passing limited credentials to each agent
+
+Masters need to run with elevated permissions to be able to create new roles and permissions on the fly for each job (`AssumeRole` in AWS).
+Those limited credentials would be passed on to the agent, who would use them to talk to the external store.
+All operations would run on agents, with less load on the master, although with extra role creation operations.
+But the configuration and setup would be considerably more complex, as well as the agent side download code, requiring larger refactorings and a more complicated core API.
+This temporary role creation does not exist in all clouds nor other artifact repositories. For instance, https://docs.microsoft.com/en-us/azure/active-directory/active-directory-configurable-token-lifetimes[Azure Active Directory token lifetime] is on public preview, and in Google Cloud ACLs are not temporary.
+
+We opted for the first, simpler option.
+
+Extra care needs to be taken so agents do not have any access to the blob store.
+
+In the common case where the vm instances are assigned roles (`IAM role` in AWS) the instance where the master runs should have access to the blob store but the agents should run in a different instance where its role does not allow it.
+
+In a Kubernetes environment this means either using different node pools for masters or agents or using something like https://github.com/jtblin/kube2iam[kube2iam] to have different roles per pod.
+
+
+== Infrastructure Requirements
+
+Ideally we could use Jenkins infrastructure to do live testing with S3, which is not currently possible due to lack of AWS account.
```

Comment:
Would be nice to briefly describe the fallback infrastructure you are going to use

Reply:
done